src/client/datascience/context/telemetry.ts

```diff
@@ -0,0 +1,139 @@
+    } else if (isErrorType(error, JupyterKernelPromiseFailedError)) {
+        return 'kernelpromisetimeout';
+    } else if (isErrorType(error, IpyKernelNotInstalledError)) {
+        return 'kernelpromisetimeout';
+    } else if (isErrorType(error, CancellationError)) {
+        return 'cancelled';
+    } else if (isErrorType(error, JupyterSessionStartError)) {
+        return 'jupytersession';
+    } else if (isErrorType(error, JupyterConnectError)) {
+        return 'jupyterconnection';
+    } else if (isErrorType(error, JupyterSelfCertsError)) {
+        return 'jupyterselfcert';
+    } else if (isErrorType(error, JupyterInstallError)) {
+        return 'jupyterinstall';
+    } else if (isErrorType(error, KernelDiedError)) {
+        return 'kerneldied';
+    }
+    return 'unknown';
+}
+export function sendKernelTelemetryEvent<P extends IEventNamePropertyMapping, E extends keyof P>(
+    resource: Resource,
+    eventName: E,
+    durationMs?: Record<string, number> | number,
+    properties?: P[E],
+    ex?: Error
+) {
+    const addOnTelemetry = getContextualPropsForTelemetry(resource);
+    if (addOnTelemetry) {
+        const props = properties || {};
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        sendTelemetryEvent(eventName as any, durationMs, Object.assign(props, addOnTelemetry), ex);
+    } else {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        sendTelemetryEvent(eventName as any, durationMs, properties, ex);
+    }
+}
+export function sendKernelTelemetryWhenDone<P extends IEventNamePropertyMapping, E extends keyof P>(
+    resource: Resource,
+    eventName: E,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    promise: Promise<any> | Thenable<any>,
+    stopWatch?: StopWatch,
+    properties?: P[E]
+) {
+    const addOnTelemetry = getContextualPropsForTelemetry(resource);
+    if (addOnTelemetry) {
+        const props = properties || {};
+        sendTelemetryWhenDone(
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            eventName as any,
+            promise,
+            stopWatch,
+            Object.assign(props, addOnTelemetry),
+            true
+        );
+    } else {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        sendTelemetryWhenDone(eventName as any, promise, stopWatch, properties, true);
+    }
+}
+export function trackResourceInformation(resource: Resource, information: Partial<ContextualTelemetryProps>) {
+    if (!resource) {
+        return;
+    }
+    const key = getUriKey(resource);
+    const currentData: ResourceSpecificTelemetryProperties = trackedInfo.get(key) || {
+        resourceType: getResourceType(resource)
+    };
+    if (information.kernelConnection) {
+        const interpreter = information.kernelConnection.interpreter;
+        if (interpreter) {
+            currentData.pythonEnvironmentType = interpreter.envType;
```

Comment:
Yeah this looks like we only track interpreter once.

Reply:
Yes, basically we track/add the information & when we send the data, we send what we have.

Reply:
Could we add a counter to the number of interpreters? Actually no, why not add a list of interpreters?

Reply:
> Could we add a counter to the number of interpreters? Actually no, why not add a list of interpreters?


Will add counter, not sure the list will be of any use.

Reply:
Added a counter for number of interpreters as well.